### PR TITLE
fix: improve zap logging capabilities

### PR DESCRIPTION
**Pull request by grahambrereton-form3**
Truncate log fields over a certain limit.

Configure structured logging and conventional field names for timestamp, level, etc.

Allow configuration of log level and max field size through environment.

## Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`

## Cherry-pick to release branches (optional)

> This PR should be cherry-picked to the following release branches:

- [ ] release-2.6
- [ ] release-2.5

### CHANGELOG

> Must include at least one of them.

- [ ] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

### Tests

> Must include at least one of them.

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test

### Side effects

- [ ] **Breaking backward compatibility**

## DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

**Comment by grahambrereton-form3** [comment]
Example log output with the limit set to 1024 bytes for testing:
```json
{"level":"info","@timestamp":"2023-08-24T17:03:42.643312116Z","logger":"desiredphase","caller":"desiredphase/controller.go:129","message":"modify desiredPhase","desiredPhase":"Run"}
{"level":"info","@timestamp":"2023-08-24T17:03:42.643321913Z","logger":"pipeline","caller":"pipeline/pipeline.go:78","message":"reconcile result","result":{"Requeue":false,"RequeueAfter":299356690221}}
{"level":"info","@timestamp":"2023-08-24T17:03:42.643343255Z","logger":"pipeline","caller":"pipeline/pipeline.go:78","message":"reconcile result","result":{"Requeue":false,"RequeueAfter":0}}
{"level":"info","@timestamp":"2023-08-24T17:03:42.643360401Z","logger":"records","caller":"records/controller.go:118","message":"iterating record","record":{"id":"nats-gke/nats-1","selectorKey":".","phase":"Not Injected/Wait","injectedCount":0,"recoveredCount":0},"desiredPhase":"Run"}
{"level":"info","@timestamp":"2023-08-24T17:03:42.643372752Z","logger":"records","caller":"records/controller.go:149","message":"apply chaos","id":"nats-gke/nats-1"}
{"level":"info","@timestamp":"2023-08-24T17:03:42.643405468Z","logger":"partition","caller":"partition/impl.go:60","message":"partition Apply","chaos":"TRUNCATED {\"kind\":\"NetworkChaos\",\"apiVersion\":\"chaos-mesh.org/v1alpha1\",\"metadata\":{\"name\":\"partition-nats-gke\",\"namespace\":\"nats-gke\",\"uid\":\"10249711-16dd-4027-96f5-6f2a1fd402ce\",\"resourceVersion\":\"106565\",\"generation\":5,\"creationTimestamp\":\"2023-08-24T17:03:42Z\",\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"chaos-mesh.org/v1alpha1\\\",\\\"kind\\\":\\\"NetworkChaos\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"name\\\":\\\"partition-nats-gke\\\",\\\"namespace\\\":\\\"nats-gke\\\"},\\\"spec\\\":{\\\"action\\\":\\\"part"}
{"level":"info","@timestamp":"2023-08-24T17:03:42.643478814Z","logger":"records","caller":"records/controller.go:218","message":"updating records","records":[{"id":"nats-gke/nats-1","selectorKey":".","phase":"Injected","injectedCount":1,"recoveredCount":0,"events":[{"type":"Succeeded","operation":"Apply","timestamp":"2023-08-24T17:03:42Z"}]}]}
{"level":"info","@timestamp":"2023-08-24T17:03:42.647110996Z","logger":"NetworkChaos-resource","caller":"v1alpha1/zz_generated.chaosmesh.go:1540","message":"validate update","name":"partition-nats-gke"}
{"level":"info","@timestamp":"2023-08-24T17:03:42.656408582Z","logger":"pipeline","caller":"pipeline/pipeline.go:78","message":"reconcile result","result":{"Requeue":false,"RequeueAfter":0}}
```

File: pkg/log/zap.go [hunk 1]
```diff
@@ -28,14 +31,89 @@ import (
...
+		CallerKey:      "caller",
+		FunctionKey:    zapcore.OmitKey,
+		StacktraceKey:  "stacktrace",
+		LineEnding:     zapcore.DefaultLineEnding,
+		EncodeLevel:    zapcore.LowercaseLevelEncoder,
+		EncodeTime:     zapcore.RFC3339NanoTimeEncoder,
+		EncodeDuration: zapcore.SecondsDurationEncoder,
+		EncodeCaller:   zapcore.ShortCallerEncoder,
+		NewReflectedEncoder: func(w io.Writer) zapcore.ReflectedEncoder {
+			enc := json.NewEncoder(newTruncatingWriter(w, maxFieldSize))
```
Comment (andrew-bulford-form3) [inline pkg/log/zap.go:68]: To reduce the change in behaviour, potentially the truncating writer could only be used if `LOG_MAX_FIELD_SIZE` is set - if the environment variable isn't found the default writer could be used.

**Review by andrew-bulford-form3 — APPROVED** [review]
This looks great, thank you for solving this!

I think this would be great to get merged so we can resolve the logging problems we're seeing, but if the ambition is to upstream it I wonder if the maintainers would be happy with the default logger having this extra functionality baked in?  Possibly they might prefer for the truncating writer to be implemented in a separate package and then the the logger constructed using that writer in places where it's needed (i.e. maybe it's only the `chaos-controller-manager` that has this problem), rather than this logger being used by default everywhere.  Not to say we shouldn't get this merged, just flagging it as something to consider if you're intending to submit this to the upstream repo.

**Comment by grahambrereton-form3** [comment]
@andrew-bulford-form3 you're right this would be hard to merge back upstream. The initial implementation was just meant to be something robust enough for us to use in Form3.

The good news is that this is now deployed in in our dev stacks, so we shouldn't see the logging issues we were seeing when the network chaos experiments run and the chaos daemon is stuck in pending for the applicable nodes.

I'm going to take another look at this and see what configuration would make sense. I think your suggestion of  having a config that by default disables the log truncation unless a max size value is provided makes a lot of sense for that aspect. There's also the other logging config changes that support Form3/logz.io conventions, like `@timestamp` for the timestamp field that might not be universal, so will have to look at the best way to inject the config and provide sensible defaults before raising a PR back to upstream.

**Comment by andrew-bulford-form3** [comment]
@grahambrereton-form3 that all sounds good, thanks again for getting this fixed so quickly!  Definitely worth just getting something that works for us for now any worrying about the OSS contribution later :ok_hand:.

**Pull request closed without merging**